### PR TITLE
daemon,client,cmd/snap: improve output after snap install/refresh/remove

### PR DESCRIPTION
**Pull request by mvo5**
This is a strawman branch for bug https://bugs.launchpad.net/snappy/+bug/1574830 (and the micro task in the trello board).

[ Needs https://github.com/ubuntu-core/snappy/pull/1093 ]

Show "Done" after snap removal and the `snap list <pkg>` output after
install/refresh.

Closes: #1574830

### CODE REVIEW

File: cmd/snap/cmd_snap_op_test.go [hunk 1]
```diff
@@ -185,8 +201,8 @@ func (s *SnapSuite) TestInstallPathDevMode(c *check.C) {
 	rest, err := snap.Parser().ParseArgs([]string{"install", "--devmode", snapPath})
 	c.Assert(err, check.IsNil)
 	c.Assert(rest, check.DeepEquals, []string{})
-	c.Check(s.Stdout(), check.Equals, "")
+	//c.Check(s.Stdout(), check.Equals, "")
```
Comment (zyga) [inline cmd/snap/cmd_snap_op_test.go]: I'd add a TODO to do something about stdout here

File: cmd/snap/cmd_list.go [hunk 2]
```diff
@@ -46,10 +47,11 @@ func (s snapsByName) Len() int           { return len(s) }
 func (s snapsByName) Less(i, j int) bool { return s[i].Name < s[j].Name }
 func (s snapsByName) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
 
-func (cmdList) Execute([]string) error {
+func (cmdList) Execute(args []string) error {
 	cli := Client()
 	filter := client.SnapFilter{
 		Sources: []string{"local"},
+		Query:   strings.Join(args, ","),
```
Comment (zyga) [inline cmd/snap/cmd_list.go]: If we set the query why isn't that showing in tests?

**Comment by mvo5** [comment]
As discussed on tg:
- when removal finishes we replace the progress/spinner line with a  `Done.`
- snap install/refresh will do a `snap list <the-pkg-that-just-got-installed-or-refreshed>`
- `snap list` will also display the snap revision in a new `rev` column

File: cmd/snap/cmd_list.go (hunk 2 above)
Comment (mvo5) [inline cmd/snap/cmd_list.go]: @zyga what tests exactly ;)

Comment (mvo5) [inline cmd/snap/cmd_list.go]: @zyga you raise a very good point, this is currently not tested so I added tests for cmd_list.go. Thank you!

**Comment by mvo5** [comment]
This is implementing the suggestions outlined above now. Sorry that it is a bit longer than expected, I also had to refactor/update some of the tests for the new output.

File: cmd/snap/cmd_list_test.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"gopkg.in/check.v1"
+
+	snap "github.com/ubuntu-core/snappy/cmd/snap"
+)
+
+func (s *SnapSuite) TestList(c *check.C) {
```
Comment (niemeyer) [inline cmd/snap/cmd_list_test.go]: Thank you!

File: cmd/snap/cmd_list_test.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+		default:
+			c.Fatalf("expected to get 1 requests, now on %d", n+1)
+		}
+
+		n++
+	})
+	rest, err := snap.Parser().ParseArgs([]string{"list"})
+	c.Assert(err, check.IsNil)
+	c.Assert(rest, check.DeepEquals, []string{})
+	//c.Check(s.Stdout(), check.Equals, "")
```
Comment (niemeyer) [inline cmd/snap/cmd_list_test.go]: Why are these commented out?  Can we please either remove if it shouldn't be here, or explain if it's something temporary that we need to fix?

File: cmd/snap/cmd_list_test.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+		default:
+			c.Fatalf("expected to get 1 requests, now on %d", n+1)
+		}
+
+		n++
+	})
+	rest, err := snap.Parser().ParseArgs([]string{"list", "foo"})
+	c.Assert(err, check.IsNil)
+	c.Assert(rest, check.DeepEquals, []string{})
+	//c.Check(s.Stdout(), check.Equals, "")
```
Comment (niemeyer) [inline cmd/snap/cmd_list_test.go]: Same as above.

File: cmd/snap/cmd_snap_op.go [hunk 6]
```diff
@@ -115,7 +115,11 @@ func (x *cmdRemove) Execute([]string) error {
 		return err
 	}
 
-	return wait(cli, changeID)
+	if err := wait(cli, changeID); err != nil {
+		return err
+	}
+	fmt.Println("Done")
```
Comment (niemeyer) [inline cmd/snap/cmd_snap_op.go]: Can we please Fpritnln into the local Stdout, which sends the line to tests and prevents spurious output when running go test?

File: cmd/snap/cmd_snap_op.go [hunk 7]
```diff
@@ -142,7 +146,19 @@ func (x *cmdInstall) Execute([]string) error {
 		return err
 	}
 
-	return wait(cli, changeID)
+	if err := wait(cli, changeID); err != nil {
```
Comment (niemeyer) [inline cmd/snap/cmd_snap_op.go]: Should we change wait so it returns the last change, so we don't need to re-query below?

File: cmd/snap/cmd_snap_op.go (hunk 6 above)
Comment (niemeyer) [inline cmd/snap/cmd_snap_op.go]: Actually, we can probably drop this line now that we agreed to have the list output.

File: cmd/snap/cmd_list_test.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
```
Comment (pedronis) [inline cmd/snap/cmd_list_test.go]: this needs now a build tag:

// +build !integrationcoverage

(there's run-check test about this)

File: cmd/snap/cmd_snap_op.go [hunk 9]
```diff
@@ -142,7 +146,19 @@ func (x *cmdInstall) Execute([]string) error {
...
+		return err
+	}
+
+	// extract the snapName from the change, important for sideloaded
```
Comment (pedronis) [inline cmd/snap/cmd_snap_op.go]: we know from above if it's a sideload or not, feels we should error more if it's a sideload and we don't get actually get a name

File: cmd/snap/cmd_snap_op.go (hunk 6 above)
Comment (pedronis) [inline cmd/snap/cmd_snap_op.go]: this is for Remove afaiu

**Comment by mvo5** [comment]
Thanks for this excellent review. I addressed the points mentioned. Please let me know if there is anything else I can do.

**Comment by niemeyer** [comment]
retest this please

**Pull request merged**